### PR TITLE
[GP-137] - Opt in to assets

### PR DESCRIPTION
**Pull request by esyeng**
Ensure users are opted into GARD, GAIN, and GARDIAN on creating their first CDP
Files changed:

- transactions/ids.js
- transactions/cdp.js
- .prettierrc.json

NOTE: some unnecessary changes to spacing and line breaks were introduced via prettier - this will be fixed in later task

### CODE REVIEW

**Comment by esyeng** [comment]
@mccabe-david the changes to look at are two new functions verifyOptIn and createOptInTxn, right above openCDP, and calls for them in lines 300 - 402. need to fix the formatter as it obscures what was changed very annoyingly

File: src/transactions/cdp.js [hunk 1]
```diff
@@ -16,748 +30,865 @@ const MINRATIO = 140;
...
+        (info["amount"] -
+          devFees -
+          307000 -
+          100000 * (info["assets"].length + 1)) /
+          1000000 +
+        " Algos",
+    };
+  }
+  let optedIn = false;
+  let optedInGard = verifyOptIn(info, gardID);
```
Comment (mccabe-david) [inline src/transactions/cdp.js:301]: Could this be simplified by replacing all instances of "optedIn" with "optedInGard"? That's what optedIn referred to initially

File: src/transactions/cdp.js [hunk 2]
```diff
@@ -16,748 +30,865 @@ const MINRATIO = 140;
...
+  // Part 1: Opting in, creating needed info, etc.
+
+  const info = await infoPromise;
+  const devFees = await devFeesPromise;
+  const accountIDPromise = findOpenID(info.address);
+
+  if (
+    307000 +
+      openingMicroALGOs +
+      100000 * (info["assets"].length + 1) +
```
Comment (mccabe-david) [inline src/transactions/cdp.js:283]: This error catch should change to +3 (for 3 potential additional opt ins), and the same goes for the Maximum Deposit alert below. Ideally, we should do this alert later. See my comments on what the "start" variable represents

Comment (mccabe-david) [reply src/transactions/cdp.js:283]: Maybe this should be +4 come to think of it. I think when this was initially added we forgot to account for opt-ins altogether,

File: src/transactions/cdp.js [hunk 3]
```diff
@@ -16,748 +30,865 @@ const MINRATIO = 140;
...
+  // r2_txns.splice(3, 1)
+  let txns = r1_txns.concat(r2_txns);
+  let stxns = await signGroup(info, txns);
+  console.log(stxns);
+
+  // TODO: Adjust indexing based on including all txns
+
+  let r1_stxns = [stxns[0].blob, stxn2.blob];
+  // txn3
+  let start = 0;
```
Comment (mccabe-david) [inline src/transactions/cdp.js:400]: "start" should equal the number of tokens that need to be opted into. Furthermore, all signed txns of the form stxns[1+k].blob for k=1,2,...,start (do no pushes if start=0) should be pushed to the r1_stxns array.

**Review by mccabe-david — CHANGES_REQUESTED** [review]
Overall, the logic looks solid! Just a few edge cases to take care of. This PR will help us a lot!

**Comment by esyeng** [comment]
@mccabe-david updated to address comments

File: src/transactions/cdp.js [hunk 4]
```diff
@@ -16,748 +30,873 @@ const MINRATIO = 140;
...
+    start = 1;
+  }
+  if (!optedInGain) {
+    r1_stxns.push(stxns[3].blob);
+    start = 2;
+  }
+  if (VERSION != "TESTNET1" && !optedInGardian) {
+    r1_stxns.push(stxns[4].blob);
+    start = 3;
+  }
```
Comment (mccabe-david) [inline src/transactions/cdp.js:410]: There's still an issue here for users who have opted-in to GARD and GAIN, but not GARDIAN

**Review by mccabe-david — APPROVED** [review]
I think everything should work now, but I'll run a few tests before merging

File: src/transactions/cdp.js [hunk 5]
```diff
@@ -399,16 +397,16 @@ export async function openCDP(openingALGOs, openingGARD) {
...
-    start = 2;
+    r1_stxns.push(stxns[2+start].blob);
+    start += 1;
   }
-  if (VERSION != "TESTNET1" && !optedInGardian) {
-    r1_stxns.push(stxns[4].blob);
-    start = 3;
+  if (!optedInGardian) {
+    r1_stxns.push(stxns[2+start].blob);
+    start += 1;
```
Comment (esyeng) [inline src/transactions/cdp.js:409]: ah yeah this makes more sense, thanks for the adjustment

**Pull request merged**